tools/collect_hana_sr_diag.sh

```diff
@@ -0,0 +1,645 @@
+        log_message "INFO" "Collecting: firewall-cmd --list-all"
+        if check_and_run_remote_command "$1" "firewall-cmd" "sudo firewall-cmd --list-all" "$2"; then
+            check_file "$2"
+        else
+            log_message "ERROR" "Failed to collect firewall rules on $1."
+        fi
+    else
+        log_message "INFO" "firewalld is not running on $1. Skipping firewall rule collection."
+        echo "firewalld is not running" > "$2"
+    fi
+}
+
+collect_system_info() {
+    log_message "INFO" "Collecting OS version"
+    if check_and_run_remote_command "$1" "cat" "cat /etc/os-release" "$2/os_version.txt"; then
+        check_file "$2/os_version.txt"
+    fi
+
+    log_message "INFO" "Collecting Pacemaker version"
+    if check_and_run_remote_command "$1" "rpm" "rpm -q pacemaker" "$2/pacemaker_version.txt"; then
+        check_file "$2/pacemaker_version.txt"
+    fi
+
+    log_message "INFO" "Collecting crmsh version"
+    if check_and_run_remote_command "$1" "rpm" "rpm -q crmsh" "$2/crmsh_version.txt"; then
+        check_file "$2/crmsh_version.txt"
+    fi
+
+    log_message "INFO" "Collecting package details"
+    if check_and_run_remote_command "$1" "rpm" "rpm -qa | grep -E 'pacemaker|crmsh|resource.*agent|fence.*agent' | xargs -r rpm -qi" "$2/packages.txt"; then
+        check_file "$2/packages.txt"
+    fi
+}
+
+collect_traces() {
+    local node_ip=$1
+    local output_dir=$2
+    local trace_pattern=$3
+    local output_file="$output_dir/${trace_pattern}_trace.log"
+
+    log_message "INFO" "Collecting ${trace_pattern} traces"
+    local remote_command="sudo find /usr/sap/${SAP_SID}/HDB${INSTANCE_NUM}/ -name '${trace_pattern}_*.trc' -print0 | while IFS= read -r -d '' file; do echo -e \"\n\n=== \${file} ===\n\"; sudo tail -n 200 \"\${file}\"; done"
+    if check_and_run_remote_command "${node_ip}" "find" "${remote_command}" "${output_file}"; then
+        check_file "${output_file}"
+    fi
+}
+
+collect_listening_ports() {
+    local node_ip=$1
+    local output_dir=$2
+    local tool_name="ss"
+    local remote_command="sudo ss -tlnp"
+    local output_file="${output_dir}/ss_tlnp.txt"
+
+    log_message "INFO" "Collecting: Listening Ports (ss -tlnp)"
+    if check_and_run_remote_command "${node_ip}" "${tool_name}" "${remote_command}" "${output_file}" "${LOG_FILE}"; then
+        check_file "${output_file}"
+        return $? # Return the status of check_file
+    else
+        log_message "ERROR" "Failed to collect listening ports on ${node_ip}."
+        return 1
+    fi
+}
+
+detect_hana_replication_port() {
+    local node_ip=$1
+    local output_dir=$2
+    local ss_output_file="${output_dir}/ss_tlnp.txt"
+    local detected_port=""
+
+    if [[ -f "$ss_output_file" ]]; then
+        # Try to find an external IP for hdbnameserver first, exclude 127.0.0.1 and 0.0.0.0
+        detected_port=$(grep 'hdbnameserver' "$ss_output_file" | grep -v '127.0.0.1' | grep -v '0.0.0.0' | grep -oP ':\K\d+' | head -n 1)
```

Comment:
```suggestion
        detected_port=$(grep 'hdbnameserver' "$ss_output_file" | grep -v -E '127.0.0.1|0.0.0.0' | grep -oP ':\K\d+' | head -n 1)
```
And perhaps the other `grep` commands can also be combined, but not sure how to do it without looking at the contents of `$ss_output_file`